streaming/src/subtle/aes_ctr_hmac.rs

```diff
@@ -0,0 +1,300 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+//! AES-CTR-HMAC based implementation of the [`tink::StreamingAead`] trait.
+
+use super::noncebased;
+use aes_ctr::cipher::stream::{Key, NewStreamCipher, SyncStreamCipher};
+use std::convert::TryInto;
+use tink::{proto::HashType, subtle::random::get_random_bytes, utils::wrap_err, Mac, TinkError};
+
+/// The size of the nonces used as IVs for CTR.
+pub const AES_CTR_HMAC_NONCE_SIZE_IN_BYTES: usize = 16;
+
+/// The size of the nonce prefix.
+pub const AES_CTR_HMAC_NONCE_PREFIX_SIZE_IN_BYTES: usize = 7;
+
+/// The size of the HMAC key.
+pub const AES_CTR_HMAC_KEY_SIZE_IN_BYTES: usize = 32;
+
+#[derive(Clone)]
+enum AesCtrKeyVariant {
+    Aes128(Key<aes_ctr::Aes128Ctr>),
+    Aes256(Key<aes_ctr::Aes256Ctr>),
+}
+
+/// `AesCtrHmac` implements streaming AEAD encryption using AES-CTR and HMAC.
+///
+/// Each ciphertext uses new AES-CTR and HMAC keys. These keys are derived using
+/// HKDF and are derived from the key derivation key, a randomly chosen salt of
+/// the same size as the key and a nonce prefix.
+#[derive(Clone)]
+pub struct AesCtrHmac {
+    pub main_key: Vec<u8>,
+    hkdf_alg: HashType,
+    key_size_in_bytes: usize,
+    tag_alg: HashType,
+    tag_size_in_bytes: usize,
+    ciphertext_segment_size: usize,
+    plaintext_segment_size: usize,
+    first_ciphertext_segment_offset: usize,
+}
+
+impl AesCtrHmac {
+    /// Initialize an AES_CTR_HMAC primitive with a key derivation key and encryption parameters.
+    ///
+    /// `main_key` is input keying material used to derive sub keys.
+    /// `hkdf_alg` is a MAC algorithm hash type, used for the HKDF key derivation.
+    /// `key_size_in_bytes` is the key size of the sub keys.
+    /// `tag_alg` is the MAC algorithm hash type, used for generating per segment tags.
+    /// `tag_size_in_bytes` is the size of the per segment tags.
+    /// `ciphertext_segment_size` is the size of ciphertext segments.
+    /// `first_segment_offset` is the offset of the first ciphertext segment.
+    pub fn new(
+        main_key: &[u8],
+        hkdf_alg: HashType,
+        key_size_in_bytes: usize,
+        tag_alg: HashType,
+        tag_size_in_bytes: usize,
+        ciphertext_segment_size: usize,
+        first_segment_offset: usize,
+    ) -> Result<AesCtrHmac, TinkError> {
+        if main_key.len() < 16 || main_key.len() < key_size_in_bytes {
```

Comment:
So it's fine for main_key to be longer than key_size_in_bytes? If so, will sub keys be truncated? Perhaps worth a comment?

Reply:
Added a comment.